File: Algorithm/C++/958. Check Completeness of a Binary Tree.cc
/*
 * Created on Sun Dec 16 2018 15:5:50
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


#include <vector>

using std::vector;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// check level by level
class Solution {
 public:
  bool isCompleteTree(TreeNode* root) {
    vector<TreeNode*> curr{root};
    bool missed = false;
    while(curr.size() > 0) {
      vector<TreeNode*> next;
      for(auto node : curr) {
        if ((missed && (node->left != NULL || node->right != NULL)) || (node->left == NULL && node->right != NULL)) {
          return false;
        }
        if (node->left)
          next.push_back(node->left);
        else
          missed = true;
        if (node->right)
          next.push_back(node->right);
        else
          missed = true;   
      }
      curr = next;
    }
    return true;
  }
};

File: Algorithm/Python/958. Check Completeness of a Binary Tree.py
# -*- coding: utf-8 -*-
# Created on Sun Dec 16 2018 10:45:13
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# check level by level
class Solution(object):
    def isCompleteTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        curr = [root]
        missed = False
        while curr:
            next = []
            for node in curr:
                if (missed and (node.left or node.right)) or (node.left == None and node.right != None):
                    return False
                if node.left:
                    next.append(node.left)
                else:
                    missed = True
                if node.right:
                    next.append(node.right)
                else:
                    missed = True
            curr = next
        return True    